layer/VkLayer_FROG_gamescope_wsi.cpp

```diff
@@ -948,23 +1042,25 @@ namespace GamescopeWSILayer {
         }
       }
 
-      VkResult result = pDispatch->QueuePresentKHR(queue, pPresentInfo);
+      VkResult result = pDispatch->QueuePresentKHR(queue, &presentInfo);
 
-      for (uint32_t i = 0; i < pPresentInfo->swapchainCount; i++) {
-        VkSwapchainKHR swapchain = pPresentInfo->pSwapchains[i];
+      for (uint32_t i = 0; i < presentInfo.swapchainCount; i++) {
+        VkSwapchainKHR swapchain = presentInfo.pSwapchains[i];
 
         auto UpdateSwapchainResult = [&](VkResult newResult) {
-          if (pPresentInfo->pResults && pPresentInfo->pResults[i] >= VK_SUCCESS)
-            pPresentInfo->pResults[i] = newResult;
+          if (presentInfo.pResults && presentInfo.pResults[i] >= VK_SUCCESS)
+            presentInfo.pResults[i] = newResult;
           if (result >= VK_SUCCESS)
             result = newResult;
         };
 
         if (auto gamescopeSwapchain = GamescopeSwapchain::get(swapchain)) {
-
-          if ((limiterOverride == 1 && gamescopeSwapchain->presentMode != VK_PRESENT_MODE_FIFO_KHR) ||
-              (limiterOverride != 1 && gamescopeSwapchain->presentMode != gamescopeSwapchain->originalPresentMode)) {
-              fprintf(stderr, "[Gamescope WSI] Forcing swapchain recreation as frame limiter changed.\n");
+          // If we are a frame limiter aware application like DXVK or VKD3D-Proton, we don't
+          // transparently change their vkQueuePresent to just FIFO modes, we change what is
+          // exposed as supported in order for them to handle presentation latency like they
+          // would as in FIFO mode.
+          if (frameLimiterAware && forceFifo && gamescopeSwapchain->presentMode != VK_PRESENT_MODE_FIFO_KHR) {
```

Comment:
Shouldn't this be in the opposite direction as well? If forceFifo is turned off, that should trigger recreation, right? Probably needs to be checked against previous value of forceFifo, not the current present mode. Application might want to use non-FIFO, but currently cannot until OUT_OF_DATE_KHR is signalled (so it can re-query supported present modes, etc).

Reply:
If vkd3d-proton is only going to requery present modes after recreation and not-per present then that would be required.

If that's the easiest path then I can re-add the code that does that on this path.

Reply:
Querying per-present is very awkward.